subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalJavaCompilationIntegrationTest.groovy

```diff
@@ -26,48 +27,204 @@ abstract class AbstractCrossTaskIncrementalJavaCompilationIntegrationTest extend
+        impl.snapshot { run language.compileTaskName }
+
+        when:
+        source api: ["class B { final static $constantType x = $newValue; /* change */ void blah() { /* avoid flakiness by changing compiled file length*/ } }"]
+        run "impl:${language.compileTaskName}"
+
+        then:
+        impl.recompiledClasses('ImplA')
+
+        where:
+        constantType | constantValue   | newValue
+        'boolean'    | 'false'         | 'true'
+        'byte'       | '(byte) 125'    | '(byte) 126'
+        'short'      | '(short) 666'   | '(short) 555'
+        'int'        | '55542'         | '444'
+        'long'       | '5L'            | '689L'
+        'float'      | '6f'            | '6.5f'
+        'double'     | '7d'            | '7.2d'
+        'String'     | '"foo"'         | '"bar"'
+        'String'     | '"foo" + "bar"' | '"bar"'
+    }
+
+    @Unroll
+    def "change in an upstream class with non-private constant causes rebuild if constant is referenced in field (#constantType)"() {
+        source api: ["class A {}", "class B { final static $constantType x = $constantValue; }"], impl: ["class ImplA extends A { final $constantType foo = B.x; }", "class ImplB {int foo() { return 2; }}"]
+        impl.snapshot { run language.compileTaskName }
+
+        when:
+        source api: ["class B { final static $constantType x = $newValue; /* change */ void blah() { /* avoid flakiness by changing compiled file length*/ } }"]
         run "impl:${language.compileTaskName}"
 
         then:
-        // impl.recompiledClasses('ImplA') //  Can re-enable with compiler plugins. See gradle/gradle#1474
-        impl.recompiledClasses('ImplA', 'ImplB')
+        impl.recompiledClasses('ImplA')
 
         where:
-        constantType | constantValue
-        'boolean'    | 'false'
-        'byte'       | '(byte) 125'
-        'short'      | '(short) 666'
-        'int'        | '55542'
-        'long'       | '5L'
-        'float'      | '6f'
-        'double'     | '7d'
-        'String'     | '"foo"'
-        'String'     | '"foo" + "bar"'
-    }
-
-    // This test describes the current behavior, not the intended one, expressed by the test just above
-    def "changing an unused non-private constant causes full rebuild"() {
+        constantType | constantValue   | newValue
+        'int'        | '55542'         | '444'
+        'String'     | '"foo" + "bar"' | '"bar"'
+    }
+
+    def "non-abi change to constant origin class does not causes compilation"() {
+        source impl: ["class A { final static int x = 1; int method() { return 1; } }", "class B { int method() { return A.x; }  }"]
+        impl.snapshot { run language.compileTaskName }
+
+        when:
+        source impl: ["class A { final static int x = 1; int method() { return 2; } }"]
+        run language.compileTaskName
+
+        then:
+        impl.recompiledClasses('A', 'B')
```

Comment:
:x: This test does not do what it says. The name of the test says that 'B' should not be recompiled. But then it asserts that 'B' is recompiled.

Reply:
Moved this test to JavaSourceIncrementalCompilationIntegrationTest and add comment to t.